accepted/2022/enhance-vector-codegen.md

```diff
@@ -0,0 +1,679 @@
+# Upgrade `Vector` API SIMD Acceleration Capability in .NET
+
+**Owner** [Anthony Canino](https://github.com/anthonycanino) 
+
+.NET support for SIMD acceleration through `Vector<T>` API allows developers to transparently harness the power of advanced SIMD ISAs without expert-level knowledge of low-level hardware details and optimization techniques. Given its variable-length nature, `Vector<T>` allows to adapt to the most performant SIMD ISA implementation available for the platform. However, internal .NET libraries use `Vector<T>` as a fallback SIMD code path, where hand-optimized intrinsic code paths, e.g, SSE or AVX2, provide more performance due to `Vector<T>` platform-agnostic API. 
+
+This is problematic for a number of reasons:
+
+  1. As new SIMD ISAs release, .NET developers must write additional SIMD code paths *per optimized library*, increasing code complexity and maintenance burden. At best, much time has to be dedicated to these libraries; at worst, libraries do not get upgraded for latest hardware advancements.
+
+  2. Hand-optimized intrinsics lock software into a specific ISA, often with fixed thresholds for determining when to select a SIMD accelerated codepath for performance gains. In a JIT environment where we can glean performance characteristics of the underlying platform, intrinsics prevent adapting to the most _performant_ SIMD ISA available for a given _workload_.
+
+In this design document, we propose to extend `Vector<T>` to serve as a vessel for frictionless SIMD adoption, both internal to .NET libraries, and to external .NET developers. As a realization of this goal, we propose the following:
+
+  1. Upgrading `Vector<T>` to serve as a sufficiently powerful interface for writing both internal hardware accelerated libraries and external developer code. 
```

Comment:
I have less experience with the source generators as a concept, so when I wrote this, I envisioned the JIT took care of this.

In the spirit of the document, would a source generator approach require the developer to "regenerate" their code as future ISAs become available/implemented, or are the source generators able to be tightly integrated into the pipeline that it's not necessarily a burden for the developer to do so?

Reply:
"source generators" in the Roslyn sense are components that plug directly into the compiler; they're handed all the information the compiler has about the source in the project and are able to add code to the compilation unit.  A key limitation is they can't replace code, only add code, at least today (and the expectation is that even if some replacement is enabled in the future, it'll be very constrained).  One of the most common forms of this is writing a partial class or method, often with an attribute applied to it, and the generator then fills in the implementation.  You can see this, for example, with the JsonSerialization generator that shipped in .NET 6, where a developer just writes a partial class attributed in a certain way, and the generator emits into it all of the logic for serializing the relevant types, or the new LibraryImport and RegexGenerators in .NET 7, where the developer writes a partial method and the generator fills in the implementation of that method.

Reply:
[`Compute](https://github.com/Sergio0694/ComputeSharp) by @Sergio0694 does something that is conceptually similar in that it takes C# code and generates code that can run on the GPU.

I had also prototyped a very basic source generator a while back which handles directive driven vectorization.

For `vectorized` code we need a software fallback for when vectorization isn't supported so it's feasible for a dev to do something like:
```csharp
public static partial int Sum(ReadOnlySpan<int> values);

[Vectorize(nameof(Sum), ...)]
private static int Sum_SoftwareFallback(ReadOnlySpan<int> values)
{
    // ...
}
```

The generator can then process `Sum_SoftwareFallback` and provide an implementation of the public `Sum` method powered by `Vector64/128/256/512<T>` or `Vector<T>`.

It would be quite a bit of work to enable, but would have some benefits over typical "auto-vectorization" approaches. An analyzer that looks for potentially vectorizable code and suggests a refactoring to put it in the above shape would also be feasible to help suggest to users where vectorization is feasible.

Reply:
In this sense, we are really talking about source-level auto-vectorization as opposed to JIT-level template generation, correct?

Reply:
Source level `directive-driven vectorization` (rather than `auto-vectorization`). It could still be "template driven" (rather than relying on analyzing scalar code patterns) if we felt that was the best approach and there are multiple options/possibilities (particularly when viewed in combination with the IL trimmer/linker).

That is, in all cases it requires some level of "user opt-in" (such as via a `Vectorize` attribute the generator recognizes). Noting that we could do some form assistance in here in recognizing vectorizable patterns and suggesting to the user they add the `Vectorize` attribute.

Whether we then do the vectorization based on some `Vector<T>` template (including generating a scalar path) or do recognition of a scalar algorithm and convert it to `Vector64/128/256/512<T>` or `Vector<T>` code is something that we could do either of.